vowpalwabbit/core/src/reductions/epsilon_decay.cc

```diff
@@ -33,8 +35,129 @@ void epsilon_decay_score::update_bounds(float w, float r)
+  // Process each model, then update the upper/lower bounds for each model
+  for (int64_t i = 0; i < num_models; ++i)
+  {
+    if (!_constant_epsilon)
+    { ep_fts.epsilon = VW::reductions::epsilon_decay::decayed_epsilon(_scored_configs[i][i].update_count); }
+    if (!base.learn_returns_prediction) { base.predict(examples, _weight_indices[i]); }
+    base.learn(examples, _weight_indices[i]);
+    for (const auto& a_s : examples[0]->pred.a_s)
+    {
+      if (a_s.action == labelled_action)
+      {
+        const float w = (logged.probability > 0) ? a_s.score / logged.probability : 0;
+        for (int64_t j = 0; j <= i; ++j) { _scored_configs[i][j].update_bounds(w, r); }
+        break;
+      }
+    }
+  }
+}
+
+// Promote model and all those lower with distance swap_dist
+void epsilon_decay_data::promote_model(int64_t model_ind, int64_t swap_dist)
+{
+  for (; model_ind >= 0; --model_ind)
+  {
+    for (int64_t score_ind = 0; score_ind < model_ind + 1; ++score_ind)
+    {
+      _scored_configs[model_ind + swap_dist][score_ind + swap_dist] = std::move(_scored_configs[model_ind][score_ind]);
+    }
+    std::swap(_weight_indices[model_ind + swap_dist], _weight_indices[model_ind]);
+  }
+}
+
+// Rebalance greater models to match lower shifted models
+void epsilon_decay_data::rebalance_greater_models(int64_t model_ind, int64_t swap_dist, int64_t num_models)
+{
+  int64_t greater_model = model_ind + swap_dist + 1;
+  for (int64_t curr_mod = greater_model; curr_mod < num_models; ++curr_mod)
+  {
+    for (int64_t score_ind = greater_model - swap_dist; score_ind > 0; --score_ind)
+    { _scored_configs[curr_mod][score_ind] = std::move(_scored_configs[curr_mod][score_ind - swap_dist]); }
+  }
+}
+
+// Clear values in removed weights and scores
+void epsilon_decay_data::clear_weights_and_scores(int64_t swap_dist, int64_t num_models)
+{
+  uint64_t params_per_weight = 1;
+  while (params_per_weight < static_cast<uint64_t>(num_models)) { params_per_weight *= 2; }
+  for (int64_t model_ind = 0; model_ind < num_models; ++model_ind)
+  {
+    for (int64_t score_ind = 0;
+         score_ind < std::min(static_cast<int64_t>(_scored_configs[model_ind].size()), swap_dist); ++score_ind)
+    { _scored_configs[model_ind][score_ind].reset_stats(_epsilon_decay_alpha, _epsilon_decay_tau); }
+  }
+  for (int64_t ind = 0; ind < swap_dist; ++ind) { _weights.clear_offset(_weight_indices[ind], params_per_weight); }
+}
+
+void epsilon_decay_data::shift_model(int64_t model_ind, int64_t swap_dist, int64_t num_models)
+{
+  promote_model(model_ind, swap_dist);
+  rebalance_greater_models(model_ind, swap_dist, num_models);
+  clear_weights_and_scores(swap_dist, num_models);
+}
+
+void epsilon_decay_data::check_score_bounds()
+{
+  // If the lower bound of a model exceeds the upperbound of the champion, migrate the new model as
+  // the new champion.
+  auto num_models = static_cast<int64_t>(_scored_configs.size());
```

Comment:
you should make a variable.. something like
`size_t final_model_idx = _scored_configs.size() - 1;`
which would let you get rid of all this `num_models -1` business